install/cupy_builder/_command.py

```diff
@@ -110,7 +125,17 @@ def build_extension(self, ext: setuptools.Extension) -> None:
         ctx = cupy_builder.get_context()
 
         # Compile "*.cu" files into object files.
-        sources_cpp, extra_objects = compile_device_code(ctx, ext)
+        sources_cpp, extra_objects, futures = compile_device_code(ctx, ext)
+
+        # We may have many TUs in the queue waiting for nvcc to complete them,
+        # so for safety we should wait for the completion here
+        done, _ = concurrent.futures.wait(
```

Comment:
`build_extension()` will be called in parallel (`CUPY_NUM_BUILD_JOBS` threads), so this will cause `CUPY_NUM_BUILD_JOBS * CUPY_NUM_BUILD_JOBS * CUPY_NUM_NVCC_THREADS` threads to be launched at maximum. It's difficult for users to control this parallelism. How about doing nvcc compilation in `build_extensions()` (which is called before starting extension build in parallel) instead of here?

Reply:
Well, technically it's true, but in practice when we start building the `thrust`/`cub` modules all other modules are done, so when I monitored the progress it's much less (https://github.com/cupy/cupy/pull/6941#issuecomment-1200775453). That said, moving the `nvcc` part to `build_extensions()` sounds like a good idea, as it increases the parallelism (compiling all *.cu at once).

Reply:
Done in 8173f5c.